src/core/thread/router_table.cpp

```diff
@@ -116,11 +116,20 @@ bool RouterTable::IsAllocated(uint8_t aRouterId) const
 void RouterTable::UpdateAllocation(void)
 {
     uint8_t indexMap[Mle::kMaxRouterId + 1];
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    uint8_t minRouterIdAllowed = 0, maxRouterIdAllowed = Mle::kMaxRouterId;
+
+    GetAllowedRouterIdRange(&minRouterIdAllowed, &maxRouterIdAllowed);
+#endif
 
     mActiveRouterCount = 0;
 
     // build index map
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    for (uint8_t routerId = minRouterIdAllowed; routerId <= maxRouterIdAllowed; routerId++)
```

Comment:
I think UpdateAllocation method should not be changed. `mActiveRouterCount` calculation should not be impacted by the Router ID range. 

Router indexes in `mRouters` should also be unchanged.